lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -2535,6 +2575,269 @@ void StaticRepeatOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
+  if (isa<RegisterOp>(operation))
+    ret = operation->getResult(1);
+  else if (isa<MemoryOp, DivSPipeLibOp, DivUPipeLibOp, MultPipeLibOp,
+               RemSPipeLibOp, RemUPipeLibOp>(operation)) {
+    // Get the value of the writer_en port or go port.
+    ret = operation->getResult(2);
+  } else if (isa<InstanceOp>(operation)) {
+    // Get the go port of the instance through the "go" attribute
+    InstanceOp instanceOp = cast<InstanceOp>(operation);
+    auto portInfo = instanceOp.getReferencedComponent().getPortInfo();
+    for (size_t i = 0; i != portInfo.size(); i++) {
+      if (portInfo[i].hasAttribute("go"))
+        ret = operation->getResult(i);
+    }
+  } else if (isa<PrimitiveOp>(operation)) {
+    // Get the go part of the primitive through the "calyx.go" attribute
+    PrimitiveOp primOp = cast<PrimitiveOp>(operation);
+    auto moduleExternOp = primOp.getReferencedPrimitive();
+    auto argAttrs = moduleExternOp.getArgAttrsAttr();
+    for (size_t i = 0; i != argAttrs.size(); i++)
+      if (DictionaryAttr dictAttr = dyn_cast<DictionaryAttr>(argAttrs[i]))
+        if (!dictAttr.empty())
+          if (dictAttr.begin()->getName().getValue() == "calyx.go")
+            ret = primOp.getResult(i);
+  }
+
+  return ret;
+}
+
+// Get the value of the done port within an instance.
+Value InvokeOp::getInstDoneValue() {
+  ComponentOp componentOp = (*this)->getParentOfType<ComponentOp>();
+  Operation *operation = componentOp.lookupSymbol(getCallee());
+  Value ret = nullptr;
+  // The done port of these instances is the last result, so the ssa value
+  // of the last result is taken.
+  if (isa<RegisterOp, MemoryOp, DivSPipeLibOp, DivUPipeLibOp, MultPipeLibOp,
+          RemSPipeLibOp, RemUPipeLibOp>(operation)) {
+    size_t doneIdx = operation->getResults().size() - 1;
+    ret = operation->getResult(doneIdx);
+  } else if (isa<InstanceOp>(operation)) {
+    //  Get the go port of the instance through the "done" attribute
+    InstanceOp instanceOp = cast<InstanceOp>(operation);
+    auto portInfo = instanceOp.getReferencedComponent().getPortInfo();
+    for (size_t i = 0; i != portInfo.size(); i++) {
+      if (portInfo[i].hasAttribute("done"))
+        ret = operation->getResult(i);
+    }
+  } else if (isa<calyx::PrimitiveOp>(operation)) {
+    // Get the go part of the primitive through the "calyx.done" attribute
+    PrimitiveOp primOp = cast<PrimitiveOp>(operation);
+    auto moduleExternOp = primOp.getReferencedPrimitive();
+    auto resAttrs = moduleExternOp.getResAttrsAttr();
+    for (size_t i = 0; i != resAttrs.size(); i++)
+      if (DictionaryAttr dictAttr = dyn_cast<DictionaryAttr>(resAttrs[i]))
+        if (!dictAttr.empty())
+          if (dictAttr.begin()->getName().getValue() == "calyx.done")
+            ret = primOp.getResult(i);
+  }
+  return ret;
+}
+
+// A helper function that gets the number of go or done ports in
+// hw.module.extern.
+static size_t getHwModuleExtPortNumber(mlir::ArrayAttr attrsAttr,
```

Comment:
If this is specifically for a hw.module.extern, then why doesn't it take a `HWModuleExternOp` as an argument?
Secondly, the comment to this function doesn't really match with what the function actually does - that is, taking a `StringRef string` and then checking for that string - is it because the `string` is the identifier for `go` or `done`? better naming and docs here :).